File: benchmark/crosslang/adapters/php/sdk.php
<?php

declare(strict_types=1);

/**
 * The php SDK BASELINE — raw PDO + hand-SQL for benchmark_* (the fair 1.0x denominator).
 *
 * The php twin of adapters/py/sdk.py + the go SDK cell (adapters/go/sdk.go): reads rows DIRECTLY by
 * column name (NOT via the litedbmodel runtime — that is the ir cell). All placeholders are PDO
 * positional `?`; the child IN-clause is an `IN (?, …)` list on ALL dialects (PDO_pgsql binds an
 * `IN (?,…)` list fine, so no pg array-bind needed). `filterPaginateSort` binds `published` = true on
 * postgres (a boolean column — pg is strict) and 1 on sqlite/mysql. mysql has NO RETURNING, so upsert +
 * tx-chain inserts do an explicit re-select (mirror the seam behavior, not the code). v1-faithful
 * returning (upsert → [{id}], no-returning writes → null, tx → {committed,state}).
 */

require_once __DIR__ . '/canon.php';
require_once __DIR__ . '/db.php';

/** @return list<string> */
function bench_batch_emails(): array
{
    return array_map(static fn ($i) => "many{$i}@bench.com", range(0, 9));
}

/** @return list<string> */
function bench_batch_names(): array
{
    return array_map(static fn ($i) => "Many {$i}", range(0, 9));
}

/** @return list<string> */
function bench_upsert_emails(): array
{
    return array_merge(['user1@example.com', 'user2@example.com'], array_slice(bench_batch_emails(), 0, 8));
}

// ── reads ──
function sdk_find_all(RawDb $db): string
{
    return canon_rows($db->query('SELECT id, email, name FROM benchmark_users ORDER BY id ASC LIMIT 100'), BENCH_FIELDS['findAll']);
}

function sdk_find_first(RawDb $db): string
{
    return canon_rows($db->query('SELECT id, email, name FROM benchmark_users WHERE name LIKE ? LIMIT 1', ['User%']), BENCH_FIELDS['findFirst']);
}

function sdk_find_unique(RawDb $db): string
{
    return canon_rows($db->query('SELECT id, email, name FROM benchmark_users WHERE email = ? LIMIT 1', ['user500@example.com']), BENCH_FIELDS['findUnique']);
}

function sdk_filter_paginate_sort(RawDb $db): string
{
    // `published` is BOOLEAN on pg (rejects int=bool), int on sqlite/mysql — bind per dialect.
    $published = $db->dialect === 'postgres' ? true : 1;
    $rows = $db->query('SELECT id, title, content, published, author_id, created_at FROM benchmark_posts WHERE published = ? ORDER BY created_at DESC LIMIT 20 OFFSET 10', [$published]);
    return canon_rows($rows, BENCH_FIELDS['filterPaginateSort']);
}

// ── single writes (v1: no-returning → null; upsert → [{id}]) ──
function sdk_create(RawDb $db): string
{
    $db->execute('INSERT INTO benchmark_users (email, name) VALUES (?, ?)', ['new@bench.com', 'New']);
    return 'null';
}

function sdk_update(RawDb $db): string
{
    $db->execute('UPDATE benchmark_users SET name = ? WHERE id = ?', ['Updated 100', 100]);
    return 'null';
}

function sdk_upsert(RawDb $db): string
{
    if ($db->dialect === 'mysql') {
        $db->execute('INSERT INTO benchmark_users (email, name) VALUES (?, ?) ON DUPLICATE KEY UPDATE email = VALUES(email), name = VALUES(name)', ['user1@example.com', 'Upserted One']);
        $rows = $db->query('SELECT id FROM benchmark_users WHERE email = ? ORDER BY id', ['user1@example.com']);
    } else {
        $rows = $db->query('INSERT INTO benchmark_users (email, name) VALUES (?, ?) ON CONFLICT (email) DO UPDATE SET email = excluded.email, name = excluded.name RETURNING id', ['user1@example.com', 'Upserted One']);
    }
    return canon_rows($rows, BENCH_FIELDS['upsert']);
}

// ── batch writes (v1: no-returning → null) ──
/**
 * @param list<string> $emails
 * @param list<string> $names
 */
function sdk_insert_many_values(RawDb $db, array $emails, array $names, string $tail): string
{
    $tuples = implode(', ', array_fill(0, count($emails), '(?, ?)'));
    $params = [];
    foreach ($emails as $i => $e) {
        $params[] = $e;
        $params[] = $names[$i];
    }
    $db->execute("INSERT INTO benchmark_users (email, name) VALUES {$tuples}{$tail}", $params);
    return 'null';
}

function sdk_create_many(RawDb $db): string
{
    return sdk_insert_many_values($db, bench_batch_emails(), bench_batch_names(), '');
}

function sdk_upsert_many(RawDb $db): string
{
    $tail = $db->dialect === 'mysql'
        ? ' ON DUPLICATE KEY UPDATE email = VALUES(email), name = VALUES(name)'
        : ' ON CONFLICT (email) DO UPDATE SET email = excluded.email, name = excluded.name';
    return sdk_insert_many_values($db, bench_upsert_emails(), bench_batch_names(), $tail);
}

function sdk_update_many(RawDb $db): string
{
    // hand-OPTIMIZED single CASE update (not a per-row loop); NO returning (v1) → null.
    $names = bench_batch_names();
    $cases = [];
    for ($i = 0; $i < 10; $i++) {
        $cases[] = 'WHEN ' . ($i + 1) . ' THEN ?';
    }
    $db->execute('UPDATE benchmark_users SET name = CASE id ' . implode(' ', $cases) . ' END WHERE id IN (1,2,3,4,5,6,7,8,9,10)', $names);
    return 'null';
}

// ── read + relation (parent + ONE batched IN child, N+1 avoided) ──
/**
 * @param list<mixed> $keys
 * @return array{0:string,1:list<mixed>}
 */
function sdk_in_clause(array $keys): array
{
    $marks = implode(', ', array_fill(0, count($keys), '?'));
    return ["IN ({$marks})", array_values($keys)];
}

/**
 * @param list<mixed> $parentParams
 * @param list<string> $parentFields
 * @param list<string> $childFields
 */
function sdk_rel_single(RawDb $db, string $parentSql, array $parentParams, string $parentKey, array $parentFields, string $childSqlTmpl, string $childKey, array $childFields, string $rel): string
{
    $parents = $db->query($parentSql, $parentParams);
    $keys = array_map(static fn ($r) => $r[$parentKey], $parents);
    [$inClause, $childParams] = sdk_in_clause($keys);
    $children = $db->query(str_replace('{IN}', $inClause, $childSqlTmpl), $childParams);
    $groups = [];
    foreach ($children as $c) {
        $groups[$c[$childKey]][] = $c;
    }
    $ps = [];
    $cs = [];
    foreach ($parents as $r) {
        $ps[] = canon_row($r, $parentFields);
        $kids = $groups[$r[$parentKey]] ?? [];
        $cs[] = '[' . implode(',', array_map(static fn ($c) => canon_row($c, $childFields), $kids)) . ']';
    }
    return rel_json($rel, $ps, $cs);
}

function sdk_nested_find_all(RawDb $db): string
{
    $m = BENCH_REL_FIELDS['nestedFindAll'];
    return sdk_rel_single($db, 'SELECT id, email, name FROM benchmark_users ORDER BY id ASC LIMIT 100', [], 'id', $m['parent'],
        'SELECT id, title, author_id FROM benchmark_posts WHERE author_id {IN} ORDER BY id ASC', 'author_id', $m['child'], $m['rel']);
}

function sdk_nested_find_first(RawDb $db): string
{
    $m = BENCH_REL_FIELDS['nestedFindFirst'];
    return sdk_rel_single($db, 'SELECT id, email, name FROM benchmark_users WHERE name LIKE ? LIMIT 1', ['User%'], 'id', $m['parent'],
        'SELECT id, title, author_id FROM benchmark_posts WHERE author_id {IN} ORDER BY id ASC', 'author_id', $m['child'], $m['rel']);
}

function sdk_nested_find_unique(RawDb $db): string
{
    $m = BENCH_REL_FIELDS['nestedFindUnique'];
    return sdk_rel_single($db, 'SELECT id, email, name FROM benchmark_users WHERE email = ? LIMIT 1', ['user1@example.com'], 'id', $m['parent'],
        'SELECT id, title, author_id FROM benchmark_posts WHERE author_id {IN} ORDER BY id ASC', 'author_id', $m['child'], $m['rel']);
}

// FULL 3-level SDK (#119): THREE batched queries (parents; level-2 by parent key; level-3 by ALL
// level-2 ids) + client stitch into {rows,posts,comments} (comments flattened per parent, in post order).
// Single-key stitch (nestedRelations); the composite-key variant is bespoke below.
/**
 * @param list<array<string,mixed>> $parents
 */
function sdk_stitch3(RawDb $db, string $op, array $parents, string $parentKey, string $postsSqlTmpl, string $postsKey, string $postsId, string $commentsSqlTmpl, string $commentsKey): string
{
    $m = BENCH_REL3_FIELDS[$op];
    $pkeys = array_map(static fn ($r) => $r[$parentKey], $parents);
    [$pIn, $pParams] = sdk_in_clause($pkeys);
    $posts = $pkeys ? $db->query(str_replace('{IN}', $pIn, $postsSqlTmpl), $pParams) : [];
    $pids = array_map(static fn ($r) => $r[$postsId], $posts);
    [$cIn, $cParams] = sdk_in_clause($pids);
    $comments = $pids ? $db->query(str_replace('{IN}', $cIn, $commentsSqlTmpl), $cParams) : [];
    $postsByParent = [];
    foreach ($posts as $p) {
        $postsByParent[$p[$postsKey]][] = $p;
    }
    $commentsByPost = [];
    foreach ($comments as $c) {
        $commentsByPost[$c[$commentsKey]][] = $c;
    }
    $rowsS = [];
    $postsS = [];
    $commentsS = [];
    foreach ($parents as $r) {
        $ps = $postsByParent[$r[$parentKey]] ?? [];
        $rowsS[] = canon_row($r, $m['parent']);
        $postsS[] = canon_rows($ps, $m['posts']);
        $flat = [];
        foreach ($ps as $p) {
            foreach ($commentsByPost[$p[$postsId]] ?? [] as $c) {
                $flat[] = $c;
            }
        }
        $commentsS[] = canon_rows($flat, $m['comments']);
    }
    return rel3_json($rowsS, $postsS, $commentsS);
}

function sdk_nested_relations(RawDb $db): string
{
    $users = $db->query('SELECT id, email, name FROM benchmark_users ORDER BY id ASC LIMIT 100');
    return sdk_stitch3($db, 'nestedRelations', $users, 'id',
        'SELECT id, title, author_id FROM benchmark_posts WHERE author_id {IN} ORDER BY id ASC', 'author_id', 'id',
        'SELECT id, body, post_id FROM benchmark_comments WHERE post_id {IN} ORDER BY id ASC', 'post_id');
}

function sdk_composite_relations(RawDb $db): string
{
    // tenant_id fixed = 1: level-2 posts + level-3 comments both filtered by tenant, stitched by sub-key.
    $m = BENCH_REL3_FIELDS['compositeRelations'];
    $tusers = $db->query('SELECT tenant_id, user_id, name FROM benchmark_tenant_users WHERE tenant_id = ? ORDER BY user_id ASC', [1]);
    $tposts = $db->query('SELECT tenant_id, post_id, user_id, title FROM benchmark_tenant_posts WHERE tenant_id = ? ORDER BY post_id ASC', [1]);
    $tcomments = $db->query('SELECT tenant_id, comment_id, post_id, body FROM benchmark_tenant_comments WHERE tenant_id = ? ORDER BY comment_id ASC', [1]);
    $postsByUser = [];
    foreach ($tposts as $p) {
        $postsByUser[$p['user_id']][] = $p;
    }
    $commentsByPost = [];
    foreach ($tcomments as $c) {
        $commentsByPost[$c['post_id']][] = $c;
    }
    $rowsS = [];
    $postsS = [];
    $commentsS = [];
    foreach ($tusers as $u) {
        $ps = $postsByUser[$u['user_id']] ?? [];
        $rowsS[] = canon_row($u, $m['parent']);
        $postsS[] = canon_rows($ps, $m['posts']);
        $flat = [];
        foreach ($ps as $p) {
            foreach ($commentsByPost[$p['post_id']] ?? [] as $c) {
                $flat[] = $c;
            }
        }
        $commentsS[] = canon_rows($flat, $m['comments']);
    }
    return rel3_json($rowsS, $postsS, $commentsS);
}

// ── transactions (BEGIN … COMMIT/ROLLBACK, then {committed,state}) ──
function sdk_insert_user_id(RawDb $db, string $email, string $name): int
{
    if ($db->dialect === 'mysql') {
        $db->execute('INSERT INTO benchmark_users (email, name) VALUES (?, ?)', [$email, $name]);
        return (int) $db->pdo->lastInsertId();
    }
    $rows = $db->query('INSERT INTO benchmark_users (email, name) VALUES (?, ?) RETURNING id', [$email, $name]);
    return (int) $rows[0]['id'];
}

function sdk_upsert_user_id(RawDb $db, string $email, string $name): int
{
    if ($db->dialect === 'mysql') {
        $db->execute('INSERT INTO benchmark_users (email, name) VALUES (?, ?) ON DUPLICATE KEY UPDATE email = VALUES(email), name = VALUES(name)', [$email, $name]);
        $rows = $db->query('SELECT id FROM benchmark_users WHERE email = ? ORDER BY id', [$email]);
        return (int) $rows[0]['id'];
    }
    $rows = $db->query('INSERT INTO benchmark_users (email, name) VALUES (?, ?) ON CONFLICT (email) DO UPDATE SET email = excluded.email, name = excluded.name RETURNING id', [$email, $name]);
    return (int) $rows[0]['id'];
}

/** @return array{0:list<array<string,mixed>>,1:list<array<string,mixed>>} */
function sdk_state_snapshot(RawDb $db): array
{
    $users = $db->query('SELECT id, email, name FROM benchmark_users ORDER BY id');
    $posts = $db->query('SELECT id, title, author_id FROM benchmark_posts ORDER BY id');
    return [$users, $posts];
}

function sdk_run_tx(RawDb $db, callable $body): string
{
    $committed = false;
    $db->begin();
    try {
        $body();
        $db->commit();
        $committed = true;
    } catch (\Throwable) {
        $db->rollback();
        $committed = false;
    }
    [$users, $posts] = sdk_state_snapshot($db);
    return tx_json($committed, $users, $posts);
}

function sdk_delete(RawDb $db): string
{
    return sdk_run_tx($db, function () use ($db): void {
        $uid = sdk_insert_user_id($db, 'del0@bench.com', 'Del');
        $db->execute('DELETE FROM benchmark_users WHERE id = ?', [$uid]);
    });
}

function sdk_nested_create(RawDb $db): string
{
    return sdk_run_tx($db, function () use ($db): void {
        $uid = sdk_insert_user_id($db, 'nc@bench.com', 'NC');
        $db->execute('INSERT INTO benchmark_posts (author_id, title) VALUES (?, ?)', [$uid, 'NC Post']);
    });
}

function sdk_nested_update(RawDb $db): string
{
    return sdk_run_tx($db, function () use ($db): void {
        $db->execute('UPDATE benchmark_users SET name = ? WHERE id = ?', ['NU', 7]);
        $db->execute('UPDATE benchmark_posts SET title = ? WHERE author_id = ?', ['NU Post', 7]);
    });
}

function sdk_nested_upsert(RawDb $db): string
{
    return sdk_run_tx($db, function () use ($db): void {
        $uid = sdk_upsert_user_id($db, 'user1@example.com', 'NUp');
        $db->execute('INSERT INTO benchmark_posts (author_id, title) VALUES (?, ?)', [$uid, 'NUp Post']);
    });
}

/** @return array<string,callable(RawDb):string> */
function sdk_dispatch(): array
{
    return [
        'findAll' => 'sdk_find_all', 'filterPaginateSort' => 'sdk_filter_paginate_sort', 'findFirst' => 'sdk_find_first', 'findUnique' => 'sdk_find_unique',
        'create' => 'sdk_create', 'update' => 'sdk_update', 'upsert' => 'sdk_upsert',
        'createMany' => 'sdk_create_many', 'upsertMany' => 'sdk_upsert_many', 'updateMany' => 'sdk_update_many',
        'nestedFindAll' => 'sdk_nested_find_all', 'nestedFindFirst' => 'sdk_nested_find_first', 'nestedFindUnique' => 'sdk_nested_find_unique',
        'nestedRelations' => 'sdk_nested_relations', 'compositeRelations' => 'sdk_composite_relations',
        'delete' => 'sdk_delete', 'nestedCreate' => 'sdk_nested_create', 'nestedUpdate' => 'sdk_nested_update', 'nestedUpsert' => 'sdk_nested_upsert',
    ];
}

function sdk_cell(string $op, RawDb $db): string
{
    $fn = sdk_dispatch()[$op] ?? null;
    if ($fn === null) {
        throw new \InvalidArgumentException("sdk: unknown op {$op}");
    }
    return $fn($db);
}

File: benchmark/crosslang/adapters/py/sdk.py
"""The python SDK BASELINE — raw driver + hand-SQL for benchmark_* (the fair 1.0x denominator).

The python twin of the go SDK cell (adapters/go/sdk.go): reads rows DIRECTLY by column name (NOT via
the litedbmodel runtime — that is the ir cell). Dialect-aware placeholders (`?` normalized to `%s` for
pg/psycopg + mysql/pymysql by db._bind; pg arrays bind a list to `= ANY(?)`). mysql has no RETURNING,
so upsert + tx-chain inserts do an explicit strip+re-select (mirror the seam behavior, not the code).
v1-faithful returning (upsert→[{id}], no-returning writes→null, tx→{committed,state}).
"""

from __future__ import annotations

from typing import Any, Dict, List

from canon import FIELDS, REL_FIELDS, canon_rows, rel_json, tx_json
from db import RawDB

BATCH_EMAILS = [f"many{i}@bench.com" for i in range(10)]
BATCH_NAMES = [f"Many {i}" for i in range(10)]
UPSERT_EMAILS = ["user1@example.com", "user2@example.com"] + BATCH_EMAILS[:8]


# ── reads ──
def _find_all(db: RawDB) -> str:
    return canon_rows(db.query("SELECT id, email, name FROM benchmark_users ORDER BY id ASC LIMIT 100"), FIELDS["findAll"])


def _find_first(db: RawDB) -> str:
    return canon_rows(db.query("SELECT id, email, name FROM benchmark_users WHERE name LIKE ? LIMIT 1", ["User%"]), FIELDS["findFirst"])


def _find_unique(db: RawDB) -> str:
    return canon_rows(db.query("SELECT id, email, name FROM benchmark_users WHERE email = ? LIMIT 1", ["user500@example.com"]), FIELDS["findUnique"])


def _filter_paginate_sort(db: RawDB) -> str:
    # `published` is BOOLEAN on pg (psycopg rejects int=bool), int on sqlite/mysql — bind per dialect.
    published = True if db.dialect == "postgres" else 1
    rows = db.query("SELECT id, title, content, published, author_id, created_at FROM benchmark_posts WHERE published = ? ORDER BY created_at DESC LIMIT 20 OFFSET 10", [published])
    return canon_rows(rows, FIELDS["filterPaginateSort"])


# ── single writes (v1: no-returning → null; upsert → [{id}]) ──
def _create(db: RawDB) -> str:
    db.execute("INSERT INTO benchmark_users (email, name) VALUES (?, ?)", ["new@bench.com", "New"])
    return "null"


def _update(db: RawDB) -> str:
    db.execute("UPDATE benchmark_users SET name = ? WHERE id = ?", ["Updated 100", 100])
    return "null"


def _upsert(db: RawDB) -> str:
    if db.dialect == "mysql":
        db.execute("INSERT INTO benchmark_users (email, name) VALUES (?, ?) ON DUPLICATE KEY UPDATE email = VALUES(email), name = VALUES(name)", ["user1@example.com", "Upserted One"])
        rows = db.query("SELECT id FROM benchmark_users WHERE email = ? ORDER BY id", ["user1@example.com"])
    else:
        rows = db.query("INSERT INTO benchmark_users (email, name) VALUES (?, ?) ON CONFLICT (email) DO UPDATE SET email = excluded.email, name = excluded.name RETURNING id", ["user1@example.com", "Upserted One"])
    return canon_rows(rows, FIELDS["upsert"])


# ── batch writes (v1: no-returning → null) ──
def _insert_many_values(db: RawDB, emails: List[str], names: List[str], tail: str) -> str:
    tuples = ", ".join("(?, ?)" for _ in emails)
    params: List[Any] = []
    for e, n in zip(emails, names):
        params += [e, n]
    db.execute(f"INSERT INTO benchmark_users (email, name) VALUES {tuples}{tail}", params)
    return "null"


def _create_many(db: RawDB) -> str:
    return _insert_many_values(db, BATCH_EMAILS, BATCH_NAMES, "")


def _upsert_many(db: RawDB) -> str:
    tail = (
        " ON DUPLICATE KEY UPDATE email = VALUES(email), name = VALUES(name)"
        if db.dialect == "mysql"
        else " ON CONFLICT (email) DO UPDATE SET email = excluded.email, name = excluded.name"
    )
    return _insert_many_values(db, UPSERT_EMAILS, BATCH_NAMES, tail)


def _update_many(db: RawDB) -> str:
    # hand-OPTIMIZED single CASE update (not a per-row loop); NO returning (v1) → null.
    cases = " ".join(f"WHEN {i + 1} THEN ?" for i in range(10))
    db.execute(f"UPDATE benchmark_users SET name = CASE id {cases} END WHERE id IN (1,2,3,4,5,6,7,8,9,10)", list(BATCH_NAMES))
    return "null"


# ── read + relation (parent + ONE batched IN child, N+1 avoided) ──
def _in_clause(db: RawDB, keys: List[Any]):
    if db.dialect == "postgres":
        return "= ANY(?)", [list(keys)]
    marks = ", ".join("?" for _ in keys)
    return f"IN ({marks})", list(keys)


def _rel_single(db: RawDB, parent_sql, parent_params, parent_key, parent_fields, child_sql_tmpl, child_key, child_fields, rel) -> str:
    parents = db.query(parent_sql, parent_params)
    keys = [r[parent_key] for r in parents]
    in_clause, child_params = _in_clause(db, keys)
    children = db.query(child_sql_tmpl.replace("{IN}", in_clause), child_params)
    groups: Dict[Any, List[Dict[str, Any]]] = {}
    for c in children:
        groups.setdefault(c[child_key], []).append(c)
    from canon import canon_row  # local: per-row projection

    ps = [canon_row(r, parent_fields) for r in parents]
    cs = ["[" + ",".join(canon_row(c, child_fields) for c in groups.get(r[parent_key], [])) + "]" for r in parents]
    return rel_json(rel, ps, cs)


def _rel(op: str):
    m = REL_FIELDS[op]
    return m["parent"], m["child"], m["rel"]


def _nested_find_all(db: RawDB) -> str:
    p, c, rel = _rel("nestedFindAll")
    return _rel_single(db, "SELECT id, email, name FROM benchmark_users ORDER BY id ASC LIMIT 100", [], "id", p,
                       "SELECT id, title, author_id FROM benchmark_posts WHERE author_id {IN} ORDER BY id ASC", "author_id", c, rel)


def _nested_find_first(db: RawDB) -> str:
    p, c, rel = _rel("nestedFindFirst")
    return _rel_single(db, "SELECT id, email, name FROM benchmark_users WHERE name LIKE ? LIMIT 1", ["User%"], "id", p,
                       "SELECT id, title, author_id FROM benchmark_posts WHERE author_id {IN} ORDER BY id ASC", "author_id", c, rel)


def _nested_find_unique(db: RawDB) -> str:
    p, c, rel = _rel("nestedFindUnique")
    return _rel_single(db, "SELECT id, email, name FROM benchmark_users WHERE email = ? LIMIT 1", ["user1@example.com"], "id", p,
                       "SELECT id, title, author_id FROM benchmark_posts WHERE author_id {IN} ORDER BY id ASC", "author_id", c, rel)


# FULL 3-level SDK (#119): THREE batched queries (parents; level-2 by parent key; level-3 by ALL
# level-2 ids) + client stitch into {rows,posts,comments} (comments flattened per parent, in post order).
def _stitch3(db: RawDB, parents, parent_key, posts_sql_tmpl, posts_key, posts_id, comments_sql_tmpl, comments_key) -> str:
    from canon import REL3_FIELDS, canon_row, canon_rows, rel3_json

    op = "nestedRelations" if parent_key == "id" else "compositeRelations"
    m = REL3_FIELDS[op]
    pkeys = [r[parent_key] for r in parents]
    p_in, p_params = _in_clause(db, pkeys)
    posts = db.query(posts_sql_tmpl.replace("{IN}", p_in), p_params) if pkeys else []
    pids = [r[posts_id] for r in posts]
    c_in, c_params = _in_clause(db, pids)
    comments = db.query(comments_sql_tmpl.replace("{IN}", c_in), c_params) if pids else []
    posts_by_parent: Dict[Any, List[Dict[str, Any]]] = {}
    for p in posts:
        posts_by_parent.setdefault(p[posts_key], []).append(p)
    comments_by_post: Dict[Any, List[Dict[str, Any]]] = {}
    for c in comments:
        comments_by_post.setdefault(c[comments_key], []).append(c)
    rows_s, posts_s, comments_s = [], [], []
    for r in parents:
        ps = posts_by_parent.get(r[parent_key], [])
        rows_s.append(canon_row(r, m["parent"]))
        posts_s.append(canon_rows(ps, m["posts"]))
        flat_comments = [c for p in ps for c in comments_by_post.get(p[posts_id], [])]
        comments_s.append(canon_rows(flat_comments, m["comments"]))
    return rel3_json(rows_s, posts_s, comments_s)


def _nested_relations(db: RawDB) -> str:
    users = db.query("SELECT id, email, name FROM benchmark_users ORDER BY id ASC LIMIT 100")
    return _stitch3(db, users, "id",
                    "SELECT id, title, author_id FROM benchmark_posts WHERE author_id {IN} ORDER BY id ASC", "author_id", "id",
                    "SELECT id, body, post_id FROM benchmark_comments WHERE post_id {IN} ORDER BY id ASC", "post_id")


def _composite_relations(db: RawDB) -> str:
    # tenant_id fixed = 1: level-2 posts + level-3 comments both filtered by tenant, stitched by sub-key.
    tusers = db.query("SELECT tenant_id, user_id, name FROM benchmark_tenant_users WHERE tenant_id = ? ORDER BY user_id ASC", [1])
    tposts = db.query("SELECT tenant_id, post_id, user_id, title FROM benchmark_tenant_posts WHERE tenant_id = ? ORDER BY post_id ASC", [1])
    tcomments = db.query("SELECT tenant_id, comment_id, post_id, body FROM benchmark_tenant_comments WHERE tenant_id = ? ORDER BY comment_id ASC", [1])
    from canon import REL3_FIELDS, canon_row, canon_rows, rel3_json

    m = REL3_FIELDS["compositeRelations"]
    posts_by_user: Dict[Any, List[Dict[str, Any]]] = {}
    for p in tposts:
        posts_by_user.setdefault(p["user_id"], []).append(p)
    comments_by_post: Dict[Any, List[Dict[str, Any]]] = {}
    for c in tcomments:
        comments_by_post.setdefault(c["post_id"], []).append(c)
    rows_s, posts_s, comments_s = [], [], []
    for u in tusers:
        ps = posts_by_user.get(u["user_id"], [])
        rows_s.append(canon_row(u, m["parent"]))
        posts_s.append(canon_rows(ps, m["posts"]))
        comments_s.append(canon_rows([c for p in ps for c in comments_by_post.get(p["post_id"], [])], m["comments"]))
    return rel3_json(rows_s, posts_s, comments_s)


# ── transactions (BEGIN … COMMIT/ROLLBACK, then {committed,state}) ──
def _insert_user_id(db: RawDB, email: str, name: str) -> int:
    if db.dialect == "mysql":
        _, last = db.execute("INSERT INTO benchmark_users (email, name) VALUES (?, ?)", [email, name])
        return last
    rows = db.query("INSERT INTO benchmark_users (email, name) VALUES (?, ?) RETURNING id", [email, name])
    return rows[0]["id"]


def _upsert_user_id(db: RawDB, email: str, name: str) -> int:
    if db.dialect == "mysql":
        db.execute("INSERT INTO benchmark_users (email, name) VALUES (?, ?) ON DUPLICATE KEY UPDATE email = VALUES(email), name = VALUES(name)", [email, name])
        return db.query("SELECT id FROM benchmark_users WHERE email = ? ORDER BY id", [email])[0]["id"]
    rows = db.query("INSERT INTO benchmark_users (email, name) VALUES (?, ?) ON CONFLICT (email) DO UPDATE SET email = excluded.email, name = excluded.name RETURNING id", [email, name])
    return rows[0]["id"]


def _state_snapshot(db: RawDB):
    users = db.query("SELECT id, email, name FROM benchmark_users ORDER BY id")
    posts = db.query("SELECT id, title, author_id FROM benchmark_posts ORDER BY id")
    return users, posts


def _run_tx(db: RawDB, body) -> str:
    committed = False
    db.begin()
    try:
        body()
        db.commit()
        committed = True
    except Exception:
        db.rollback()
        committed = False
    users, posts = _state_snapshot(db)
    return tx_json(committed, users, posts)


def _delete(db: RawDB) -> str:
    def body():
        uid = _insert_user_id(db, "del0@bench.com", "Del")
        db.execute("DELETE FROM benchmark_users WHERE id = ?", [uid])

    return _run_tx(db, body)


def _nested_create(db: RawDB) -> str:
    def body():
        uid = _insert_user_id(db, "nc@bench.com", "NC")
        db.execute("INSERT INTO benchmark_posts (author_id, title) VALUES (?, ?)", [uid, "NC Post"])

    return _run_tx(db, body)


def _nested_update(db: RawDB) -> str:
    def body():
        db.execute("UPDATE benchmark_users SET name = ? WHERE id = ?", ["NU", 7])
        db.execute("UPDATE benchmark_posts SET title = ? WHERE author_id = ?", ["NU Post", 7])

    return _run_tx(db, body)


def _nested_upsert(db: RawDB) -> str:
    def body():
        uid = _upsert_user_id(db, "user1@example.com", "NUp")
        db.execute("INSERT INTO benchmark_posts (author_id, title) VALUES (?, ?)", [uid, "NUp Post"])

    return _run_tx(db, body)


_DISPATCH = {
    "findAll": _find_all, "filterPaginateSort": _filter_paginate_sort, "findFirst": _find_first, "findUnique": _find_unique,
    "create": _create, "update": _update, "upsert": _upsert,
    "createMany": _create_many, "upsertMany": _upsert_many, "updateMany": _update_many,
    "nestedFindAll": _nested_find_all, "nestedFindFirst": _nested_find_first, "nestedFindUnique": _nested_find_unique,
    "nestedRelations": _nested_relations, "compositeRelations": _composite_relations,
    "delete": _delete, "nestedCreate": _nested_create, "nestedUpdate": _nested_update, "nestedUpsert": _nested_upsert,
}


def sdk_cell(op: str, db: RawDB) -> str:
    fn = _DISPATCH.get(op)
    if fn is None:
        raise ValueError(f"sdk: unknown op {op}")
    return fn(db)
